File: 2594-minimum-time-to-repair-cars/2594-minimum-time-to-repair-cars.py
class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        low, high = 1, ranks[0] * cars * cars   # high = let rank[0] repair all cars
        while low < high:
            mid = low + (high - low) // 2
            if self.canRepair(ranks, cars, mid):
                high = mid
            else:
                low = mid + 1
        return low

    def canRepair(self, ranks: List[int], cars: int, maxTime: int) -> bool:
        for rank in ranks:
            repairedCarsSq = maxTime / rank
            cars -= int(repairedCarsSq ** 0.5)
            if cars <= 0: break
        return cars <= 0

File: 2594-minimum-time-to-repair-cars/2594-minimum-time-to-repair-cars.java
class Solution {
    public long repairCars(int[] ranks, int cars) {
        long low = 1, high = 1L * ranks[0] * cars * cars;   // high = let rank[0] repair all cars
        while(low < high){
            long mid = low + (high - low) / 2;
            if(canRepair(ranks, cars, mid))
                high = mid;
            else
                low = mid + 1;
        }

        return low;
    }

    private boolean canRepair(int[] ranks, int cars, long maxTime){
        for(var rank : ranks){
            var repairedCarsSq = maxTime / rank;
            cars -= (int)Math.sqrt(repairedCarsSq);
            if(cars <= 0) break;
        }

        return cars <= 0;
    }
}